File: LesBonsNumerosAPIParser.py
import re
import logging
import xml.etree.ElementTree as ET

from urllib.request import Request, urlopen

from Utils import Utils, NumbersInFrench
from Result import Result

logger = logging.getLogger("FDJ_SLACK")


class LesBonsNumerosAPIParser(object):

    API_URL = "https://www.lesbonsnumeros.com/loto/rss.xml"

    def __init__(self):
        self.string_response = None
        self.xml_node = None

    def get_data(self):

        request = Request(self.API_URL)
        self.string_response = urlopen(request).read().decode()
        # TODO check response code
        self.xml_node = ET.fromstring(self.string_response)
        logger.debug("[LesBonsNumerosAPIParser] get_data: REPONSE >> " + self.string_response)

    def parse_data(self):
        list_results = list()
        for item_idx, item in enumerate(self.xml_node[0].findall('item')):
            result_text_data = item[3].text
            result_text_data = Utils.cleanhtml(result_text_data)
            result_line_list = result_text_data.split("\n")

            date = LesBonsNumerosAPIParser.__extract_date_lottery__(result_line_list[1])
            balls = LesBonsNumerosAPIParser.__extract_balls__(result_line_list[2])
            lucky_ball = LesBonsNumerosAPIParser.__extract_lucky_ball__(result_line_list[3])
            number_winner = LesBonsNumerosAPIParser.__extract_number_winner__(result_line_list[11])
            winner_prize = LesBonsNumerosAPIParser.__extract_winner_prize__(result_line_list[11])
            next_lottery_date = LesBonsNumerosAPIParser.__extract_next_lottery_date__(result_line_list[12])
            next_lottery_prize = LesBonsNumerosAPIParser.__extract_next_lottery_prize__(result_line_list[12])

            result = Result(string_date=date,
                            balls=balls,
                            lucky_ball=lucky_ball,
                            number_winner=number_winner,
                            winner_prize=winner_prize,
                            next_lottery_string_date=next_lottery_date,
                            next_lottery_prize=next_lottery_prize)

            list_results.append(result)
        logger.debug("[LesBonsNumerosAPIParser] parse_data: LIST_RESUlTS >> " + str(list_results))
        return list_results

    @staticmethod
    def __extract_date_lottery__(line):
        return line[len("Résultats du "):]

    @staticmethod
    def __extract_balls__(line):
        line = line[len("Numéros : "):]
        return line.split(" - ")

    @staticmethod
    def __extract_lucky_ball__(line):
        return line[len("Numéro Chance : ")]

    @staticmethod
    def __extract_number_winner__(line):
        winner_number = -1
        if "Le jackpot n'a pas été remporté lors de ce tirage !" in line:
            winner_number = 0
            line = line.lower()
        elif NumbersInFrench.ONE in line:
            winner_number = 1
        elif NumbersInFrench.TWO in line:
            winner_number = 2
        elif NumbersInFrench.THREE in line:
            winner_number = 3
        elif NumbersInFrench.FOUR in line:
            winner_number = 4
        elif NumbersInFrench.FIVE in line:
            winner_number = 5
        elif NumbersInFrench.SIX in line:
            winner_number = 6
        elif NumbersInFrench.SEVEN in line:
            winner_number = 7
        elif NumbersInFrench.EIGHT in line:
            winner_number = 8
        elif NumbersInFrench.NINE in line:
            winner_number = 9
        elif NumbersInFrench.TEN in line:
            winner_number = 10
        return winner_number

    @staticmethod
    def __extract_winner_prize__(line):
        prize = re.findall(r'\d+', line.replace(" ",""))
        return 0 if len(prize) == 0 else int(prize[0])

    @staticmethod
    def __extract_next_lottery_date__(line):
        line = line[len("Le montant du jackpot du prochain tirage du "):]
        line = " ".join(line.split(" ")[:4])[:-3] # remove the 'est' attached to the year coming from the API.
        return line

    @staticmethod
    def __extract_next_lottery_prize__(line):
        # Detected numbers in the line and filter when greater than 6 characters (millions)
        return int(list(prize for prize in re.findall(r'\d+', line.replace(" ", "")) if len(prize) > 6)[0]) # too complex

File: parser/LesBonsNumerosAPIParser.go
package parser

import (
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"

	l "github.com/LaMonF/FDJ_SLACK/log"
	"github.com/LaMonF/FDJ_SLACK/model"
	"github.com/LaMonF/FDJ_SLACK/utils"

	"github.com/beevik/etree"
)

var apiURL = "https://www.lesbonsnumeros.com/loto/rss.xml"

type LesBonsNumerosAPIParser struct {
}

func NewParser() *LesBonsNumerosAPIParser {
	s := &LesBonsNumerosAPIParser{}
	return s
}

func (p *LesBonsNumerosAPIParser) GetLotteryResult() (model.LotteryResult, error) {
	var lastResult model.LotteryResult

	data, err := p.fetchData()
	if err != nil {
		return lastResult, err
	}

	results, err := p.parseData(data)
	if err != nil {
		return lastResult, err
	}

	for index, result := range results {
		if index == 0 { // only first result
			l.Info(result)
			//We can improve this post by using the URL from the POST request
			//See (https://api.slack.com/slash-commands -> Sending delayed responses)
			return result, nil
		}
	}
	return lastResult, errors.New("Last Result not found")
}

func (p *LesBonsNumerosAPIParser) fetchData() ([]byte, error) {
	l.Info("Get data from ApiURL ", apiURL)

	//Get data from URL
	response, err := http.Get(apiURL)
	if err != nil {
		return nil, fmt.Errorf("can't fetch data: %v", err)
	}

	defer response.Body.Close()
	contents, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return nil, fmt.Errorf("can't read fetched data: %v", err)
	}
	return contents, nil
}

func (p *LesBonsNumerosAPIParser) parseData(data []byte) ([]model.LotteryResult, error) {
	l.Info("Parsing data")

	//Create new XML Document to go through results
	doc := etree.NewDocument()
	if err := doc.ReadFromBytes(data); err != nil {
		return nil, err
	}

	var result []model.LotteryResult

	//For each item (using XPath to get description)
	for _, t := range doc.FindElements("//item/description") {
		l.Info("Found item")
		for _, child := range t.Child {
			if c, ok := child.(*etree.CharData); ok {

				// Manage only not empty strings
				if !utils.IsWhitespace(c.Data) {

					//Cleaning Result to remove any HTML tags (ex: <h1>)
					cleanHTML := utils.CleanHTML(c.Data)

					//Get result line by line
					resultByLine := strings.Split(cleanHTML, "\n")

					l.Info(resultByLine)

					date := extractDateLottery(resultByLine[1])
					balls, err := extractBalls(resultByLine[2])
					if err != nil {
						return nil, err
					}

					luckyBall, err := extractLuckyBall(resultByLine[3])
					if err != nil {
						return nil, err
					}
					numberWinner := 0
					winnerPrize := 0
					nextLotteryDate := "Unknown"
					nextLotteryPrize := 0
					if len(resultByLine) > 10 { // When results are not up to date len(result_line_list) < 10
						numberWinner = extractNumberWinner(resultByLine[11])
						if winnerPrize, err = extractWinnerPrize(resultByLine[11]); err != nil {
							return nil, err
						}
						nextLotteryDate = extractNextLotteryDate(resultByLine[len(resultByLine)-2])
						if nextLotteryPrize, err = extractNextLotteryPrize(resultByLine[len(resultByLine)-2]); err != nil {
							return nil, err
						}
					}

					lotteryResult := model.LotteryResult{
						Date:             date,
						Balls:            balls,
						LuckyBall:        luckyBall,
						WinnerNumber:     numberWinner,
						WinnerPrize:      winnerPrize,
						NextLotteryDate:  nextLotteryDate,
						NextLotteryPrize: nextLotteryPrize,
					}

					result = append(result, lotteryResult)
				}
			}
		}
	}

	return result, nil
}

func extractDateLottery(line string) string {
	return line[len("Résultats du "):]
}

func extractBalls(line string) ([]int, error) {
	line = line[len("Numéros : "):]
	ballsAsString := strings.Split(line, " - ")

	var balls []int
	for _, ballAsString := range ballsAsString {
		i, err := strconv.Atoi(ballAsString)
		if err != nil {
			return nil, fmt.Errorf("can't extract balls: %v", err)
		}
		balls = append(balls, i)
	}
	return balls, nil
}

func extractLuckyBall(line string) (int, error) {
	luckyBallAsString := line[len("Numéro Chance : "):]
	i, err := strconv.Atoi(luckyBallAsString)
	if err != nil {
		return 0, fmt.Errorf("can't extract lucky ball: %v", err)
	}
	return i, nil
}

func extractNumberWinner(line string) int {
	winnerNumber := -1
	line = strings.ToLower(line)
	if strings.Contains(line, "le jackpot n'a pas été remporté lors de ce tirage !") {
		winnerNumber = 0
	} else if strings.Contains(line, utils.ONE) {
		winnerNumber = 1
	} else if strings.Contains(line, utils.TWO) {
		winnerNumber = 2
	} else if strings.Contains(line, utils.THREE) {
		winnerNumber = 3
	} else if strings.Contains(line, utils.FOUR) {
		winnerNumber = 4
	} else if strings.Contains(line, utils.FIVE) {
		winnerNumber = 5
	} else if strings.Contains(line, utils.SIX) {
		winnerNumber = 6
	} else if strings.Contains(line, utils.SEVEN) {
		winnerNumber = 7
	} else if strings.Contains(line, utils.EIGHT) {
		winnerNumber = 8
	} else if strings.Contains(line, utils.NINE) {
		winnerNumber = 9
	} else if strings.Contains(line, utils.TEN) {
		winnerNumber = 10
	}
	return winnerNumber
}

// Could be done in a better way
func extractWinnerPrize(line string) (int, error) {
	indexBegin := strings.Index(line, "montant de") + len("montant de")
	if indexBegin <= len("montant de") {
		return 0, nil
	}
	indexEnd := strings.Index(line, "€")
	prizeStr := strings.Replace(line[indexBegin:indexEnd], " ", "", -1)
	prizeStr = strings.Replace(prizeStr, "&nbsp;", "", -1)
	prize, err := strconv.Atoi(prizeStr)
	if err != nil {
		return 0, fmt.Errorf("can't return winner prize: %v", err)
	}
	return prize, nil
}

func extractNextLotteryDate(line string) string {
	line = line[len("Le montant du jackpot du prochain tirage du "):]
	index := strings.Index(line, "est")
	return line[:index]
}

// Could be done in a better way
func extractNextLotteryPrize(line string) (int, error) {
	indexBegin := strings.Index(line, "est de") + len("est de")
	indexEnd := strings.Index(line, "€")
	prizeStr := strings.Replace(line[indexBegin:indexEnd], " ", "", -1)
	prize, err := strconv.Atoi(prizeStr)
	if err != nil {
		return 0, fmt.Errorf("can't extract next lottery prize: %v", err)
	}
	return prize, nil
}
